adapters/adhese/adhese.go

```diff
@@ -0,0 +1,247 @@
+			return nil, []error{err, WrapError(fmt.Sprintf("Response %v could not be parsed to JERLICIA ext.", string(response.Body)))}
+		}
+
+		if err := json.Unmarshal(response.Body, &originDataArray); err != nil {
+			return nil, []error{err, WrapError(fmt.Sprintf("Response %v could not be parsed to JERLICIA origin data.", string(response.Body)))}
+		}
+		bidResponse = convertAdheseBid(adheseBid, extArray[0], originDataArray[0])
+	} else {
+		bidResponse = convertAdheseOpenRtbBid(adheseBid)
+	}
+
+	price, _ := strconv.ParseFloat(adheseBid.Extension.Prebid.Cpm.Amount, 64)
+	width, _ := strconv.ParseUint(adheseBid.Width, 10, 64)
+	height, _ := strconv.ParseUint(adheseBid.Height, 10, 64)
+	bidResponse.Cur = adheseBid.Extension.Prebid.Cpm.Currency
+	if len(bidResponse.SeatBid) > 0 && len(bidResponse.SeatBid[0].Bid) > 0 {
+		bidResponse.SeatBid[0].Bid[0].Price = price
+		bidResponse.SeatBid[0].Bid[0].W = width
+		bidResponse.SeatBid[0].Bid[0].H = height
+	}
+
+	bidderResponse := adapters.NewBidderResponseWithBidsCapacity(5)
+
+	if len(bidResponse.SeatBid) == 0 {
+		return nil, []error{WrapError("Response resulted in an empty seatBid array.")}
+	}
+
+	var errs []error
+	for _, sb := range bidResponse.SeatBid {
+		for i := 0; i < len(sb.Bid); i++ {
+			bid := sb.Bid[i]
+			bidderResponse.Bids = append(bidderResponse.Bids, &adapters.TypedBid{
+				Bid:     &bid,
+				BidType: getBidType(bid.AdM),
+			})
+
+		}
+	}
+	return bidderResponse, errs
+}
+
+func convertAdheseBid(adheseBid AdheseBid, adheseExt AdheseExt, adheseOriginData AdheseOriginData) openrtb.BidResponse {
+	adheseExtJson, _ := json.Marshal(adheseOriginData)
+
+	return openrtb.BidResponse{
+		ID: adheseExt.Id,
+		SeatBid: []openrtb.SeatBid{{
+			Bid: []openrtb.Bid{{
+				DealID: adheseExt.OrderId,
+				CrID:   adheseExt.Id,
+				AdM:    getAdMarkup(adheseBid, adheseExt),
+				Ext:    adheseExtJson,
+			}},
+			Seat: "",
+		}},
+	}
+}
+
+func convertAdheseOpenRtbBid(adheseBid AdheseBid) openrtb.BidResponse {
+	var response openrtb.BidResponse = adheseBid.OriginData
+	if len(response.SeatBid) > 0 && len(response.SeatBid[0].Bid) > 0 {
+		response.SeatBid[0].Bid[0].AdM = adheseBid.Body
+	}
+	return response
+}
+
+func getAdMarkup(adheseBid AdheseBid, adheseExt AdheseExt) string {
+	if adheseExt.Ext == "js" {
+		if ContainsAny(adheseBid.Body, []string{"<script", "<div", "<html"}) {
+			return adheseBid.Body + "<img src='" + adheseExt.ImpressionCounter + "' style='height:1px; width:1px; margin: -1px -1px; display:none;'/>"
+		}
+		if ContainsAny(adheseBid.Body, []string{"<?xml", "<vast"}) {
+			return adheseBid.Body
+		}
+	}
+	return adheseExt.Tag
+}
+
+func getBidType(bidAdm string) openrtb_ext.BidType {
+	if bidAdm != "" && ContainsAny(bidAdm, []string{"<?xml", "<vast"}) {
+		return openrtb_ext.BidTypeVideo
+	}
+	return openrtb_ext.BidTypeBanner
+}
+
+func WrapError(errorStr string) *errortypes.BadInput {
+	return &errortypes.BadInput{Message: errorStr}
```

Comment:
See earlier comment. I don't think it's appropriate to wrap everything as `BadInput`, which signals that the publisher is making the call wrong. Some instances look better served by `BadServerResponse` in which you get a 500 back or malformed json.